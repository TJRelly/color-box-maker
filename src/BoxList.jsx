import { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);
    const addBox = (boxData) => setBoxes((boxes) => [...boxes, boxData]);
    const removeBox = (id) =>
        setBoxes((boxes) => boxes.filter((box) => box.id !== id));

    return (
        <>
            <h1>The Box Maker</h1>
            {boxes.map((box, i) => (
                <Box key={i} box={box} removeBox={removeBox} />
            ))}
            <NewBoxForm addBox={addBox} />
        </>
    );
};

export default BoxList;
