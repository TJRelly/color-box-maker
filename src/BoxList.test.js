import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
    render(<BoxList />);
});

it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function () {
    const boxList = render(<BoxList />);
    
    // no boxes yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();

    // Simulate filling out the form to add a new box
    const colorInput = boxList.getByLabelText("Color");
    const widthInput = boxList.getByLabelText("Width");
    const heightInput = boxList.getByLabelText("Height");
    const addButton = boxList.getByText("Add a new box!");

    // Change values in the form
    fireEvent.change(colorInput, { target: { value: "peachpuff" } });
    fireEvent.change(widthInput, { target: { value: "10" } });
    fireEvent.change(heightInput, { target: { value: "10" } });

    // Submit the form to add the box
    fireEvent.click(addButton);

    // expect to see a box
    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
});

it("can remove a box", function () {
    const boxList = render(<BoxList />);

    // Simulate filling out the form to add a new box
    const colorInput = boxList.getByLabelText("Color");
    const widthInput = boxList.getByLabelText("Width");
    const heightInput = boxList.getByLabelText("Height");
    const addButton = boxList.getByText("Add a new box!");

    // Change values in the form
    fireEvent.change(colorInput, { target: { value: "peachpuff" } });
    fireEvent.change(widthInput, { target: { value: "10" } });
    fireEvent.change(heightInput, { target: { value: "10" } });

    // Submit the form to add the box
    fireEvent.click(addButton);

    const removeButton = boxList.getByText("X");

    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
});
