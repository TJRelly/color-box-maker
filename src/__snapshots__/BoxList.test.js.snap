// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box List Tests matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    The Box Maker
  </h1>
  <form
    class="Box-form"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="color"
      >
        Color
      </label>
      <input
        class="form-control"
        id="color"
        name="color"
        placeholder="Enter color"
        type="string"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="width"
      >
        Width
      </label>
      <input
        class="form-control"
        id="width"
        name="width"
        placeholder="Enter width"
        type="number"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="height"
      >
        Height
      </label>
      <input
        class="form-control"
        id="height"
        name="height"
        placeholder="Enter height"
        type="number"
        value=""
      />
    </div>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Add a new box!
    </button>
  </form>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    The Box Maker
  </h1>
  <form
    class="Box-form"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="color"
      >
        Color
      </label>
      <input
        class="form-control"
        id="color"
        name="color"
        placeholder="Enter color"
        type="string"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="width"
      >
        Width
      </label>
      <input
        class="form-control"
        id="width"
        name="width"
        placeholder="Enter width"
        type="number"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="height"
      >
        Height
      </label>
      <input
        class="form-control"
        id="height"
        name="height"
        placeholder="Enter height"
        type="number"
        value=""
      />
    </div>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Add a new box!
    </button>
  </form>
</DocumentFragment>
`;
